In [13]:
import pandas as pd
import markdown
import math
import time
import numpy as np
import scipy.stats as stats
from operator import itemgetter
import pandas as pd
from collections import defaultdict
from collections import defaultdict
from timeit import Timer
from math import factorial

<p>Q0 A<br />
The code selects the tuple with the largest last element when there are tuples
with the same first element.<br />
Q0 B<br />
There are several errors in the original code. Firstly, there is an indent error.
On line5, the code's level should equal to the previous code's level.<br />
Secondly, as the range(len(sample_list)) = range(3), in line 7 and line 9, 
the index for the tuple should be less than 3.<br />
To make the code more concise, we can get over of unnecessary variable, such as
get rid of variable res. </p>

In [27]:
##Q1
def tuple01(n = int, k = int, low = int, high = int):
    list01 = np.random.randint(low, high, size=(n,k), dtype=int)
    tuple01 = [tuple(i) for i in list01]
    return tuple01
    
tuple01(4,5,0,10)

[(9, 2, 7, 6, 7), (3, 6, 8, 6, 7), (9, 6, 6, 8, 3), (1, 6, 6, 5, 5)]

In [28]:
##Q2
##A
def question2_a(sample_list, x, y):
    op = []
    for m in range(len(sample_list)):
        li = [sample_list[m]]
        for n in range(len(sample_list)):
            if (sample_list[m][x] == sample_list[n][x] and sample_list[m][y] != sample_list[n][y]):
                li.append(sample_list[n])
        op.append(sorted(li, key=lambda dd: dd[y], reverse=True)[0])
    res = list(set(op))
    return res

question2_a([(1, 3, 5), (0, 1, 2), (1, 9, 8)], 0, 2)


[(0, 1, 2), (1, 9, 8)]

In [29]:
##B
def question2_b(sample_list, x, y):
    op = []
    for m in range(len(sample_list)):
        li = [sample_list[m]]
        for n in range(len(sample_list)):
            if (sample_list[m][x] == sample_list[n][x] and sample_list[m][y] != sample_list[n][y]):
                li.append(sample_list[n])
        op.append(sorted(li, key=lambda dd: dd[y], reverse=True)[0])
    return list(set(op))

question2_a([(1, 3, 5), (0, 1, 2), (1, 9, 8)], 0, 2)


[(0, 1, 2), (1, 9, 8)]

In [30]:
##C
def question2_c(sample_list, a, b):
    d01 = defaultdict(list)
    for i in sample_list:
        d01[i[a]].append(i)
    res = [max(val, key=lambda x: x[b]) for val in d01.values()]
    return res

question2_c([(1, 3, 5), (0, 1, 2), (1, 9, 8)], 0, 2)

[(1, 9, 8), (0, 1, 2)]

In [31]:
##D
start1 = time.time() 
for i in range(1000):
    question2_a(tuple01(4,3,0,10), 0, 2) 
end1 = time.time()
time1 = end1- start1
print(time1)

start2 = time.time() 
for i in range(1000):
    question2_b(tuple01(4,3,0,10), 0, 2) 
end2 = time.time()
time2 = end2- start2
print(time2)

start3 = time.time() 
for i in range(1000):
    question2_c(tuple01(4,3,0,10), 0, 2) 
end3 = time.time()
time3 = end3- start3
print(time3)

print('In the study, each function run 1000 times with random tuple lists, and method 2a use' + str(time1) + ' minutes,and in 2b use' + str(time2) + ' minutes,and in 2c use' + str(time3) + ' minutes')


0.02436971664428711
0.020928144454956055
0.016249895095825195
In the study, each function run 1000 times with random tuple lists, and method 2a use0.02436971664428711 minutes,and in 2b use0.020928144454956055 minutes,and in 2c use0.016249895095825195 minutes


In [ ]:
##Q3
##A
data1112 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2011-2012/DEMO_G.XPT')
data1314 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2013-2014/DEMO_H.XPT')
data1516 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/DEMO_I.XPT')
data1718 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/DEMO_J.XPT')
data1 = data1112[['SEQN', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2', 'DMDMARTL', 'RIDSTATR', 'SDMVPSU', 'SDMVSTRA', 'WTMEC2YR', 'WTINT2YR']]
data1['cohort'] = '1112'
data2 = data1314[['SEQN', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2', 'DMDMARTL', 'RIDSTATR', 'SDMVPSU', 'SDMVSTRA', 'WTMEC2YR', 'WTINT2YR']]
data2['cohort'] = '1314'
data3 = data1516[['SEQN', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2', 'DMDMARTL', 'RIDSTATR', 'SDMVPSU', 'SDMVSTRA', 'WTMEC2YR', 'WTINT2YR']]
data3['cohort'] = '1516'
data4 = data1718[['SEQN', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2', 'DMDMARTL', 'RIDSTATR', 'SDMVPSU', 'SDMVSTRA', 'WTMEC2YR', 'WTINT2YR']]
data4['cohort'] = '1718'
datasum = pd.concat([data1, data2, data3, data4])
datasum.columns = ['seqn', 'ridageyr', 'ridreth3', 'dmdeduc2', 'dmdmartl', 'ridstatr', 'sdmvpsu', 'sdmvstra', 'wtmec2yr', 'wtint2yr', 'cohort']
datasum.to_pickle("./nhanes_3a.pkl")

In [33]:
##B
Oral1112 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2011-2012/OHXDEN_G.XPT')
Oral1314 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2013-2014/OHXDEN_H.XPT')
Oral1516 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/OHXDEN_I.XPT')
Oral1718 = pd.read_sas('https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/OHXDEN_J.XPT')

a= [f'OHX{i:02}TC' for i in range(1, 32)]
print(a)
b= [f'OHX{i:02}CTC' for i in range(1, 31)]
print(b)
c= [f'ohx{i:02}tc' for i in range(1, 32)]
print(a)
d= [f'ohx{i:02}ctc' for i in range(1, 31)]
print(b)
    
Oral1 = Oral1112[['SEQN', 'OHDDESTS','OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 
                      'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 
                      'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC',
                      'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 
                      'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 
                      'OHX29TC', 'OHX30TC', 'OHX31TC', 'OHX02CTC', 'OHX03CTC',
                      'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 
                      'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 
                      'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 
                      'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC',
                      'OHX28CTC', 'OHX29CTC', 'OHX30CTC']]
Oral1['cohort'] = '1112'
Oral2 = Oral1314[['SEQN', 'OHDDESTS','OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 
                      'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 
                      'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 
                      'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 
                      'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 
                      'OHX29TC', 'OHX30TC', 'OHX31TC', 'OHX02CTC', 'OHX03CTC',
                      'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 
                      'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 
                       'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 
                      'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC',
                      'OHX28CTC', 'OHX29CTC', 'OHX30CTC']]
Oral2['cohort'] = '1314'
Oral3 = Oral1516[['SEQN', 'OHDDESTS','OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 
                      'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 
                      'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 
                      'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 
                      'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 
                      'OHX29TC', 'OHX30TC', 'OHX31TC','OHX02CTC', 'OHX03CTC',
                      'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 
                      'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 
                      'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 
                      'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC',
                      'OHX28CTC', 'OHX29CTC', 'OHX30CTC']]
Oral3['cohort'] = '1516'
Oral4 = Oral1718[['SEQN', 'OHDDESTS','OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 
                      'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 
                      'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 
                      'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 
                      'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 
                      'OHX29TC', 'OHX30TC', 'OHX31TC', 'OHX02CTC', 'OHX03CTC',
                      'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 
                      'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 
                      'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 
                      'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC',
                      'OHX28CTC', 'OHX29CTC', 'OHX30CTC']]
Oral4['cohort'] = '1718'
oralsum = pd.concat([Oral1, Oral2, Oral3, Oral4])
oralsum.columns = ['seqn', 'ohddests', 'OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 
                      'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 
                      'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 
                      'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 
                      'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 
                      'OHX29TC', 'OHX30TC', 'OHX31TC', 'OHX02CTC', 'OHX03CTC',
                      'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 
                      'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 
                      'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 
                      'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC',
                      'OHX28CTC', 'OHX29CTC', 'OHX30CTC', 'cohort']
oralsum.to_pickle("./oral_3b.pkl") 

['OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC', 'OHX28TC', 'OHX29TC', 'OHX30TC', 'OHX31TC']
['OHX01CTC', 'OHX02CTC', 'OHX03CTC', 'OHX04CTC', 'OHX05CTC', 'OHX06CTC', 'OHX07CTC', 'OHX08CTC', 'OHX09CTC', 'OHX10CTC', 'OHX11CTC', 'OHX12CTC', 'OHX13CTC', 'OHX14CTC', 'OHX15CTC', 'OHX16CTC', 'OHX17CTC', 'OHX18CTC', 'OHX19CTC', 'OHX20CTC', 'OHX21CTC', 'OHX22CTC', 'OHX23CTC', 'OHX24CTC', 'OHX25CTC', 'OHX26CTC', 'OHX27CTC', 'OHX28CTC', 'OHX29CTC', 'OHX30CTC']
['OHX01TC', 'OHX02TC', 'OHX03TC', 'OHX04TC', 'OHX05TC', 'OHX06TC', 'OHX07TC', 'OHX08TC', 'OHX09TC', 'OHX10TC', 'OHX11TC', 'OHX12TC', 'OHX13TC', 'OHX14TC', 'OHX15TC', 'OHX16TC', 'OHX17TC', 'OHX18TC', 'OHX19TC', 'OHX20TC', 'OHX21TC', 'OHX22TC', 'OHX23TC', 'OHX24TC', 'OHX25TC', 'OHX26TC', 'OHX27TC',

<ipython-input-33-6ea629901ae9>:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Oral1['cohort'] = '1112'
<ipython-input-33-6ea629901ae9>:39: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Oral2['cohort'] = '1314'
<ipython-input-33-6ea629901ae9>:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#retur

In [34]:
##C
datasum.describe()
oralsum.describe()

seqn      ohddests       OHX01TC       OHX02TC       OHX03TC  \
count   35909.000000  35909.000000  34397.000000  34397.000000  34397.000000   
mean    82908.273859      1.085271      3.707562      2.885542      2.621886   
std     11815.722152      0.402831      0.719808      1.006497      0.941297   
min     62161.000000      1.000000      2.000000      2.000000      2.000000   
25%     73581.000000      1.000000      4.000000      2.000000      2.000000   
50%     83270.000000      1.000000      4.000000      2.000000      2.000000   
75%     93042.000000      1.000000      4.000000      4.000000      4.000000   
max    102956.000000      3.000000      5.000000      5.000000      5.000000   

            OHX04TC       OHX05TC       OHX06TC       OHX07TC       OHX08TC  \
count  34397.000000  34397.000000  34397.000000  34397.000000  34397.000000   
mean       2.176818      2.166352      2.023025      2.108527      2.107887   
std        0.986986      0.980418      0.851822      0.837019      0.795978   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        2.000000      2.000000      2.000000      2.000000      2.000000   
50%        2.000000      2.000000      2.000000      2.000000      2.000000   
75%        2.000000      2.000000      2.000000      2.000000      2.000000   
max        5.000000      5.000000      5.000000      5.000000      5.000000   

       ...       OHX22TC       OHX23TC       OHX24TC       OHX25TC  \
count  ...  34397.000000  34397.000000  34397.000000  34397.000000   
mean   ...      1.955054      2.015263      2.034247      2.035643   
std    ...      0.722089      0.680732      0.665893      0.668435   
min    ...      1.000000      1.000000      1.000000      1.000000   
25%    ...      2.000000      2.000000      2.000000      2.000000   
50%    ...      2.000000      2.000000      2.000000      2.000000   
75%    ...      2.000000      2.000000      2.000000      2.000000   
max    ...      5.000000      5.000000      5.000000      5.000000   

            OHX26TC       OHX27TC       OHX28TC       OHX29TC       OHX30TC  \
count  34397.000000  34397.000000  34397.000000  34397.000000  34397.000000   
mean       2.019159      1.957729      2.043928      2.113527      2.694625   
std        0.685104      0.722452      0.850334      0.927191      0.968757   
min        1.000000      1.000000      1.000000      1.000000      2.000000   
25%        2.000000      2.000000      2.000000      2.000000      2.000000   
50%        2.000000      2.000000      2.000000      2.000000      2.000000   
75%        2.000000      2.000000      2.000000      2.000000      4.000000   
max        5.000000      5.000000      5.000000      5.000000      5.000000   

            OHX31TC  
count  34397.000000  
mean       2.889031  
std        1.007436  
min        2.000000  
25%        2.000000  
50%        2.000000  
75%        4.000000  
max        5.000000  

[8 rows x 33 columns]

<p>There are  39156 cases in dataset datasum, and 35909 cases in dataset oralsum</p>